File: pwiz_tools/BiblioSpec/src/Verbosity.cpp
//
// $Id$
//
//
// Original author: Barbara Frewen <frewen@u.washington.edu>
//
// Copyright 2012 University of Washington - Seattle, WA 98195
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#include "Verbosity.h"
#include "pwiz/utility/misc/Std.hpp"

namespace BiblioSpec {

V_LEVEL Verbosity::Global_Verbosity = V_STATUS;
FILE* Verbosity::log_file = NULL;
char Verbosity::msg_buffer[1024];

void Verbosity::set_verbosity(V_LEVEL level) {
    Verbosity::Global_Verbosity = level;
    Verbosity::msg_buffer[0] = '\0';
}

void Verbosity::open_logfile(){
    Verbosity::log_file = fopen("blib-build.log", "w");
    if( Verbosity::log_file == NULL ){
        Verbosity::error("Cannot open log file 'blib-build.log'.");
    }
}

void Verbosity::close_logfile(){
    if( Verbosity::log_file != NULL ){
        fclose(log_file);
    }
}

V_LEVEL Verbosity::string_to_level(const char* level_str) {
    if( strcmp(level_str, "silent")==0) {
        return V_SILENT;
    }
    if( strcmp(level_str, "error")==0) {
        return V_ERROR;
    }
    if( strcmp(level_str, "warn")==0) {
        return V_WARN;
    }
    if( strcmp(level_str, "status")==0) {
        return V_STATUS;
    }
    if( strcmp(level_str, "debug")==0) {
        return V_DEBUG;
    }
    if( strcmp(level_str, "detail")==0) {
        return V_DETAIL;
    }
    if( strcmp(level_str, "all")==0) {
        return V_ALL;
    }

    string error_msg = "The verbosity level ";
    error_msg += level_str;
    error_msg += " is not valid";
    throw error_msg;
}

void Verbosity::error(const char* format, ...){
    va_list args;
    va_start(args, format);
    Verbosity::log(V_ERROR, format, args);
    va_end(args);
}

void Verbosity::warn(const char* format, ...){
    va_list args;
    va_start(args, format);
    Verbosity::log(V_WARN, format, args);
    va_end(args);
}

void Verbosity::status(const char* format, ...){
    va_list args;
    va_start(args, format);
    Verbosity::log(V_STATUS, format, args);
    va_end(args);
}

void Verbosity::debug(const char* format, ...){
    va_list args;
    va_start(args, format);
    Verbosity::log(V_DEBUG, format, args);
    va_end(args);
}

void Verbosity::log(V_LEVEL print_level, const char* format, va_list args)
{
    if (Verbosity::Global_Verbosity < print_level) {
        return;
    }

    // for appending to the message buffer
    char* cur_buffer_position = Verbosity::msg_buffer;
    int added = 0;

    switch (print_level) {
        case V_ERROR:
            added = sprintf(cur_buffer_position, "ERROR: ");
            cur_buffer_position += added;
            break;
        case V_WARN:
            added = sprintf(cur_buffer_position, "WARNING: ");
            cur_buffer_position += added;
            break;
        case V_STATUS:
            // No extra prefix for status.  Too ugly for command-line use.
            break;
        case V_DEBUG:
        case V_DETAIL:
        case V_ALL:
            added = sprintf(cur_buffer_position, "DEBUG: ");
            cur_buffer_position += added;
            break;
        case V_SILENT:
            break; // print nothing, shouldn't have gotten to here anyway
    }

    added = vsnprintf(cur_buffer_position, 1024 - added, format, args);
    cur_buffer_position += added;
    sprintf(cur_buffer_position, "\n");

    // if no log file, print all levels to stderr
    if (Verbosity::log_file == NULL) {
        cerr << Verbosity::msg_buffer << flush;
    } else {  // print all levels to file
        fprintf(log_file, "%s", Verbosity::msg_buffer);
        if (print_level <= V_STATUS) {
            cerr << Verbosity::msg_buffer << flush;
        }
    }

    if (print_level == V_ERROR) {

        if (Verbosity::log_file != NULL) {
            fclose(log_file);
        }
        exit(1);
    }

    Verbosity::msg_buffer[0] = '\0';
}

void Verbosity::comment(V_LEVEL print_level, 
                        const char* format, ...) {

    if(Verbosity::Global_Verbosity < print_level ) {
        return;
    }
  
    va_list args;
    va_start(args, format);
    log(print_level, format, args);
    va_end(args);
}

} // namespace

/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 4
 * End:
 */

File: pwiz-sharp/Tools/BiblioSpec/src/BiblioSpec/Verbosity.cs
// Port of pwiz_tools/BiblioSpec/src/Verbosity.{h,cpp}

using System.Diagnostics;
using System.Globalization;

namespace Pwiz.Tools.BiblioSpec;

/// <summary>
/// Verbosity / log levels, ordered from least-verbose to most-verbose. A message at level X
/// is emitted when <see cref="Verbosity.GlobalLevel"/> &gt;= X.
/// </summary>
/// <remarks>Port of <c>BiblioSpec::V_LEVEL</c>. Numeric values match the cpp enum.</remarks>
public enum VerbosityLevel
{
    /// <summary>No output (cpp: V_SILENT).</summary>
    Silent = 0,
    /// <summary>Only fatal errors (cpp: V_ERROR).</summary>
    Error = 1,
    /// <summary>Current step of execution (cpp: V_STATUS).</summary>
    Status = 2,
    /// <summary>Non-fatal warnings (cpp: V_WARN).</summary>
    Warn = 3,
    /// <summary>Extra info for debugging (cpp: V_DEBUG).</summary>
    Debug = 4,
    /// <summary>More info (cpp: V_DETAIL).</summary>
    Detail = 5,
    /// <summary>Way more output than you should ever need (cpp: V_ALL).</summary>
    All = 6,
}

/// <summary>
/// Static logging utility ported from <c>BiblioSpec::Verbosity</c>. Messages are emitted
/// to <see cref="Console.Error"/> when their level is &lt;= the current <see cref="GlobalLevel"/>.
/// <see cref="Error(string)"/> additionally throws <see cref="BlibException"/> after writing
/// (the cpp version calls <c>exit(1)</c>; throwing is the BCL-friendly equivalent and lets
/// callers in tests catch the failure).
/// </summary>
/// <remarks>
/// cpp Verbosity.cpp uses varargs printf style; in C# callers pass pre-formatted strings
/// (use string interpolation or <see cref="string.Format(System.IFormatProvider, string, object?[])"/>).
/// The cpp default is <c>V_WARN</c>; we keep the same default here.
/// </remarks>
public static class Verbosity
{
    private static readonly object _logLock = new();
    private static StreamWriter? _logFile;
    private static readonly Stopwatch _clock = Stopwatch.StartNew();

    /// <summary>Current global verbosity level. Defaults to <see cref="VerbosityLevel.Warn"/>
    /// (matching cpp <c>V_WARN</c>).</summary>
    public static VerbosityLevel GlobalLevel { get; set; } = VerbosityLevel.Warn;

    /// <summary>When true, every emitted line is prefixed with an HH:MM:SS.fff elapsed timestamp.</summary>
    public static bool TimestampEnabled { get; set; }

    /// <summary>
    /// Translate the cpp-style verbosity strings ("silent","error","warn","status","debug",
    /// "detail","all") to a <see cref="VerbosityLevel"/>. Throws <see cref="BlibException"/>
    /// on unknown input (cpp throws a bare <c>std::string</c>; we use the typed exception).
    /// </summary>
    public static VerbosityLevel StringToLevel(string levelStr)
    {
        ArgumentNullException.ThrowIfNull(levelStr);
        return levelStr switch
        {
            "silent" => VerbosityLevel.Silent,
            "error" => VerbosityLevel.Error,
            "warn" => VerbosityLevel.Warn,
            "status" => VerbosityLevel.Status,
            "debug" => VerbosityLevel.Debug,
            "detail" => VerbosityLevel.Detail,
            "all" => VerbosityLevel.All,
            _ => throw new BlibException(false, $"The verbosity level {levelStr} is not valid"),
        };
    }

    /// <summary>Open <c>blib-build.log</c> in the current directory and tee all output to it.</summary>
    public static void OpenLogfile()
    {
        lock (_logLock)
        {
            try
            {
                _logFile = new StreamWriter("blib-build.log", append: false) { AutoFlush = true };
            }
            catch (Exception ex)
            {
                throw new BlibException(false, $"Cannot open log file 'blib-build.log': {ex.Message}");
            }
        }
    }

    /// <summary>Close the log file opened by <see cref="OpenLogfile"/>, if any.</summary>
    public static void CloseLogfile()
    {
        lock (_logLock)
        {
            _logFile?.Dispose();
            _logFile = null;
        }
    }

    /// <summary>
    /// Emit a message at <see cref="VerbosityLevel.Error"/> and throw <see cref="BlibException"/>.
    /// </summary>
    /// <remarks>cpp Verbosity.cpp:175 calls <c>exit(1)</c>; throwing keeps the BCL stack
    /// intact so tests / callers can recover.</remarks>
    public static void Error(string message) => Comment(VerbosityLevel.Error, message);

    /// <summary>Emit a message at <see cref="VerbosityLevel.Warn"/>.</summary>
    public static void Warn(string message) => Comment(VerbosityLevel.Warn, message);

    /// <summary>Emit a message at <see cref="VerbosityLevel.Status"/>.</summary>
    public static void Status(string message) => Comment(VerbosityLevel.Status, message);

    /// <summary>Emit a message at <see cref="VerbosityLevel.Debug"/>.</summary>
    public static void Debug(string message) => Comment(VerbosityLevel.Debug, message);

    /// <summary>
    /// Split a (possibly multi-line) message and write each line to stderr prefixed with
    /// <c>"ERROR: "</c>. Does NOT route through <see cref="Comment"/> / <see cref="Log"/> —
    /// the per-line-prefix shape only makes sense for the per-file build-errors collected
    /// in <see cref="BlibBuilder.BuildLibrary"/>, and routing through Verbosity would
    /// either throw (Error level) or prefix only the first physical line (Warn level —
    /// the original bug that broke Skyline's regex parser in
    /// BiblioSpecLiteBuilder.IsLibraryMissingExternalSpectraError because the captured
    /// stderr lost the "Run with the -E flag" anchor on the unprefixed continuation lines).
    /// </summary>
    /// <remarks>cpp parity: BlibBuild.cpp:78 <c>WriteErrorLines</c>.</remarks>
    public static void WriteErrorLines(string message)
    {
        if (string.IsNullOrEmpty(message)) return;
        using var reader = new System.IO.StringReader(message);
        string? line;
        while ((line = reader.ReadLine()) != null)
        {
            Console.Error.WriteLine("ERROR: " + line);
        }
    }

    /// <summary>
    /// Emit a message at the given level. Suppressed when <see cref="GlobalLevel"/> &lt; level.
    /// </summary>
    public static void Comment(VerbosityLevel level, string message)
    {
        ArgumentNullException.ThrowIfNull(message);

        if (GlobalLevel < level)
        {
            // Fatal level still throws even when suppressed, matching cpp:
            // V_ERROR always calls exit(1) regardless of Global_Verbosity.
            if (level == VerbosityLevel.Error)
            {
                // AlreadyLogged=false here because we did NOT write to stderr (suppressed).
                throw new BlibException(false, message);
            }
            return;
        }

        Log(level, message);
    }

    private static void Log(VerbosityLevel level, string message)
    {
        var prefix = string.Empty;
        if (TimestampEnabled)
        {
            var elapsed = _clock.Elapsed;
            prefix = string.Format(
                CultureInfo.InvariantCulture,
                "[{0:D2}:{1:D2}:{2:D2}.{3:D3}] ",
                (int)elapsed.TotalHours,
                elapsed.Minutes,
                elapsed.Seconds,
                elapsed.Milliseconds);
        }

        var levelPrefix = level switch
        {
            VerbosityLevel.Error => "ERROR: ",
            VerbosityLevel.Warn => "WARNING: ",
            VerbosityLevel.Debug or VerbosityLevel.Detail or VerbosityLevel.All => "DEBUG: ",
            _ => string.Empty,
        };

        var line = prefix + levelPrefix + message;

        lock (_logLock)
        {
            if (_logFile is null)
            {
                Console.Error.WriteLine(line);
            }
            else
            {
                _logFile.WriteLine(line);
                if (level <= VerbosityLevel.Status)
                {
                    Console.Error.WriteLine(line);
                }
            }
        }

        if (level == VerbosityLevel.Error)
        {
            lock (_logLock)
            {
                _logFile?.Dispose();
                _logFile = null;
            }
            // AlreadyLogged=true because Log() just wrote the formatted line to stderr.
            // CLI top-level handler can use this to avoid double-printing.
            throw new BlibException(false, message) { AlreadyLogged = true };
        }
    }
}
